synapse/axon.py

```diff
@@ -1,801 +1,1195 @@
-                    if not curs.set_key(sha256):
-                        wants.append(sha256)
+    async def finish(self):
+        if self.uploader is None:
+            return
+        rv = await self.uploader.finish()
+        return rv
 
-            chan.txok(wants)
+    async def finishFile(self):
+        if self.uploader is None:
+            return
+        return await self.uploader.finishFile()
 
-    @s_glob.inpool
-    def _onAxonLocs(self, chan, mesg):
-        with chan:
-            sha256 = mesg[1].get('sha256')
-            with self.lenv.begin() as xact:
-                locs = self.getBlobLocs(xact, sha256)
-            if not locs:
-                return chan.txerr(('NoSuchFile', {}))
-            return chan.txok(locs)
+class Axon(s_cell.Cell):
 
-    @s_glob.inpool
-    def _onAxonBytes(self, chan, mesg):
+    cellapi = AxonApi
+    confdefs = (  # type: ignore
+        ('mapsize', {'type': 'int', 'defval': s_lmdb.DEFAULT_MAP_SIZE, 'doc': 'The size of the LMDB memory map'}),
+    )
 
-        with chan:
+    def __init__(self, dirn: str, conf=None) -> None:
+        s_cell.Cell.__init__(self, dirn)
 
-            chan.setq()
+        path = s_common.gendir(self.dirn, 'axon.lmdb')
+        mapsize = self.conf.get('mapsize')
+        self.lenv = lmdb.open(path, writemap=True, max_dbs=128)
+        self.lenv.set_mapsize(mapsize)
 
-            sha256 = mesg[1].get('sha256')
+        self.bloblocs = self.lenv.open_db(b'axon:blob:locs', dupsort=True, dupfixed=True) # <sha256>=blobstor_bsid
+        self.offsets = self.lenv.open_db(b'axon:blob:offsets') # <sha256>=blobstor_bsid
 
-            with self.lenv.begin() as xact:
-                locs = self.getBlobLocs(xact, sha256)
+        # Persistent settings
+        self.settings = self.lenv.open_db(b'axon:settings', dupsort=True)
 
-            if not locs:
-                return chan.txerr(('NoSuchFile', {}))
-            sess = None
-            buid = None
+        self._metrics = s_lmdb.Metrics(self.lenv)
+        self._proxykeeper = _ProxyKeeper()
 
-            for name, buid in locs:
-                sess = self.blobs.get(name)
-                if sess is not None:
-                    break
+        # Clear the global proxykeeper bsid->telepath path map (really just for unit tests)
+        _ProxyKeeper.bsidpathmap = {}
 
-            if sess is None:
-                return chan.txerr(('NotReady', {}))
+        paths = self._get_stored_blobstorpaths()
+        self.blobstorwatchers = {}
 
-            with sess.chan() as bchan:
+        # Wait a few seconds for the daemon to register all of its shared objects
+        DAEMON_DELAY = 3
+        for path in paths:
+            s_glob.plex.addLoopCoro(_run_after_delay(DAEMON_DELAY, self._start_watching_blobstor(path)))
 
-                bchan.setq()
-                bchan.tx(('blob:load', {'buid': buid}))
+        self._worker = threading.Thread(target=self._workloop, name='BlobStorWriter')
```

Comment:
in alot of cases we typically just the `@s_common.firethread` decorator.

Reply:
done